### PR TITLE
[GEOT-6605] Add SqlServer support to ImageMosaic Index

### PR DESCRIPTION
**Pull request by taba90**
JIRA ticket https://osgeo-org.atlassian.net/browse/GEOT-6605

Add  SQLServer to imagemosaic index:

- support for SqlServer factories in datastore.properties;
- added a remapping when attributes lenght is > 128 characters (SQLServer limit);
- added online tests.

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

For all pull requests:

- [x] Confirm you have read the [contribution guidelines](https://github.com/geotools/geotools/blob/master/CONTRIBUTING.md) 
- [x] You have sent a Contribution Licence Agreement (CLA) as necessary (not required for small changes, e.g., fixing typos in documentation)
- [x] Make sure the first PR targets the master branch, eventual backports will be managed later. This can be ignored if the PR is fixing an issue that only happens in a specific branch, but not in newer ones.
- [ ] The changes are not breaking the build in downstream projects using SNAPSHOT dependencies, GeoWebCache and GeoServer.

The following are required only for core and extension modules (they are welcomed, but not required, for unsupported modules):
- [x] There is an issue in [Jira](https://osgeo-org.atlassian.net/projects/GEOT) describing the bug/task/new feature (a notable exemptions is, changes not visible to end users). The ticket is for the GeoTools project, if the issue was found elsewhere it's a good practice to link to the origin ticket/issue.
- [x] The pull request contains changes related to a single objective. If multiple focuses cannot be avoided, each one is in its own commit and has a separate ticket describing it.
- [ ] PR for bug fixes and small new features are presented as a single commit
- [x] Commit message must be in the form "[GEOT-XYZW] Title of the Jira ticket"
- [x] New unit tests have been added covering the changes
- [ ] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [ ] This PR passes the [QA checks](https://docs.geotools.org/latest/developer/conventions/code/qa.html) (QA checks results will be reported by travis-ci after opening this PR)
- [ ] Documentation has been updated accordingly.

Submitting the PR does not require you to check all items, but by the time it gets merged, they should be either satisfied or inapplicable.

### CODE REVIEW

**Comment by dromagnoli** [comment]
In general, it looks good to me.

I see that the SQLServerTypeMapper has more than 90% of the code identical to the PostgisFeatureTypeMapper. Almost same constructor (beside the _index_ part), but same remapFeatureType and remap methods.
Same for OracleFeatureTypeMapper which is slightly different but shares some common pieces too.
I'm wondering if since we are introducing the 3rd wrapper with basically the same code and methods, we should think about defining a sort of BaseFeatureTypeMapper containing all the duplicated code.
What do you think?

File: modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/catalog/sqlserver/SQLServerTypeMapper.java [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+import org.opengis.feature.type.Name;
+import org.opengis.referencing.crs.CoordinateReferenceSystem;
+
+/**
+ * Specific SQLServer implementation for a {@link DataStoreWrapper} By default, SQLServer
+ * identifiers can't be longer than 128 chars. See <a
+ * href="https://docs.microsoft.com/en-us/sql/relational-databases/databases/database-identifiers?view=sql-server-ver15">SQL
+ * Syntax identifiers</a>
+ */
+public class SQLServerTypeMapper implements FeatureTypeMapper {
```
Comment (dromagnoli) [inline modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/catalog/sqlserver/SQLServerTypeMapper.java:41]: see my comment at the end of this PR about eventually moving duplicated code on a superclass.

File: modules/plugin/imagemosaic/src/test/java/org/geotools/gce/imagemosaic/ImageMosaicSQLServerIndexOnlineTest.java [hunk 2]
```diff
@@ -0,0 +1,682 @@
...
+                    + "SRID INTEGER NOT NULL,"
+                    + "TYPE VARCHAR(30) NOT NULL,"
+                    + "UNIQUE(F_TABLE_SCHEMA, F_TABLE_NAME, F_GEOMETRY_COLUMN),"
+                    + "CHECK(TYPE IN ('POINT','LINE', 'POLYGON', 'COLLECTION', 'MULTIPOINT', 'MULTILINE', 'MULTIPOLYGON', 'GEOMETRY') ))";
+
+    /**
+     * Simple Class for better testing raster manager
+     *
+     * @author Simone Giannecchini, GeoSolutions SAS
+     */
```
Comment (dromagnoli) [inline modules/plugin/imagemosaic/src/test/java/org/geotools/gce/imagemosaic/ImageMosaicSQLServerIndexOnlineTest.java:98]: Not sure you are Simone, actually  :-)

File: modules/plugin/imagemosaic/src/test/java/org/geotools/gce/imagemosaic/ImageMosaicSQLServerIndexOnlineTest.java [hunk 3]
```diff
@@ -0,0 +1,682 @@
...
+
+            // Test the output coverage
+            reader = TestUtils.getReader(timeElevURL, format);
+            elevation.setValue(Arrays.asList(NumberRange.create(0.0, 10.0)));
+            TestUtils.checkCoverage(
+                    reader,
+                    new GeneralParameterValue[] {gg, time, bkg, elevation, direct},
+                    "Time-Elevation Test");
+        } finally {
+            // dropTables(new String[] {tempFolderName1, "GEOMETRY_METADATA"});
```
Comment (dromagnoli) [inline modules/plugin/imagemosaic/src/test/java/org/geotools/gce/imagemosaic/ImageMosaicSQLServerIndexOnlineTest.java:291]: is it intentionally commented out?
needed for other type of testing?
the other tests have that line not commented.

File: modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/catalog/sqlserver/SQLServerTypeMapper.java (hunk 1 above)
Comment (taba90) [reply modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/catalog/sqlserver/SQLServerTypeMapper.java]: fixed

File: modules/plugin/imagemosaic/src/test/java/org/geotools/gce/imagemosaic/ImageMosaicSQLServerIndexOnlineTest.java (hunk 3 above)
Comment (taba90) [reply modules/plugin/imagemosaic/src/test/java/org/geotools/gce/imagemosaic/ImageMosaicSQLServerIndexOnlineTest.java]: fixed

File: modules/plugin/imagemosaic/src/test/java/org/geotools/gce/imagemosaic/ImageMosaicSQLServerIndexOnlineTest.java (hunk 2 above)
Comment (taba90) [reply modules/plugin/imagemosaic/src/test/java/org/geotools/gce/imagemosaic/ImageMosaicSQLServerIndexOnlineTest.java:98]: indeed :). Updated it

**Comment by dromagnoli** [comment]
Thanks for the updates.
+1 from me

**Pull request merged**